agent/consul/state/db/txn.go

```diff
@@ -0,0 +1,17 @@
+package db
+
+import "github.com/hashicorp/go-memdb"
+
+// ReadTxn is implemented by memdb.Txn to perform read operations.
+type ReadTxn interface {
+	Get(table, index string, args ...interface{}) (memdb.ResultIterator, error)
+	Abort()
+}
+
+// Changes wraps a memdb.Changes to include the index at which these changes
+// were made.
+type Changes struct {
+	// Index is the latest index at the time these changes were committed.
+	Index   uint64
+	Changes memdb.Changes
+}
```

Comment:
This package is new. Normally interfaces could leave in the consumer (`state`), but in this case the interfaces are used as a type in a function parameter, which is itself part of an interface. Function types don't follow the same rules as struct types, the signature has to match exactly.

This required a package separate from `state` to provide the interfaces. They could be part of `stream`, but `stream.ReadTxn` didn't make much sense. For now I've created this small package to handle the types. I suspect there are a few other things from `store` that could be part of the `db` package.

Reply:
I made a few more adjustments to the `EventPublisher` interface, which made this package unnecessary. The functions for processing changes into Events is now part of `state`, so these two types were moved to the `state` package.